subgraphs/uniswap-forks/protocols/camelot-v2/src/price.ts

```diff
@@ -0,0 +1,364 @@
+          const stable = pair.stableSwap();
+
+          let newPrice = BIGDECIMAL_ZERO;
+          if (stable) {
+            newPrice = calculateRateForStablePair(
+              pool,
+              pair,
+              token,
+              whitelistToken
+            ).times(whitelistToken.lastPriceUSD!);
+          } else {
+            newPrice = safeDiv(
+              poolAmounts.inputTokenBalances[whitelistTokenIndex],
+              poolAmounts.inputTokenBalances[priceTokenIndex]
+            ).times(whitelistToken.lastPriceUSD! as BigDecimal);
+          }
+
+          if (isValidTVL(pool, token, newPrice)) {
+            priceSoFar = newPrice;
+          }
+        }
+      }
+    }
+  }
+
+  if (priceSoFar.equals(BIGDECIMAL_ZERO)) {
+    return priceSoFar;
+  }
+
+  // Buffer token pricings that would cause large spikes on the protocol level
+  const protocol = getOrCreateProtocol();
+  const tokenTVLDelta = absBigDecimal(
+    priceSoFar
+      .times(convertTokenToDecimal(token._totalSupply, token.decimals))
+      .minus(token._totalValueLockedUSD)
+  );
+  const protocolTVLPercentageDelta = absBigDecimal(
+    safeDiv(tokenTVLDelta, protocol.totalValueLockedUSD)
+  );
+  if (protocolTVLPercentageDelta.gt(BIGDECIMAL_FIVE_PERCENT)) {
+    log.warning("Price too high for token: {} from pool: {}", [
+      token.id,
+      priceSoFar.toString(),
+    ]);
+    if (token._largeTVLImpactBuffer < PRICE_CHANGE_BUFFER_LIMIT) {
+      token._largeTVLImpactBuffer += 1;
+      token.save();
+      return token.lastPriceUSD!;
+    }
+  }
+
+  if (!token.lastPriceUSD || token.lastPriceUSD!.equals(BIGDECIMAL_ZERO)) {
+    token.save();
+    return priceSoFar;
+  }
+
+  // If priceSoFar 10x greater or less than token.lastPriceUSD, use token.lastPriceUSD
+  // Increment buffer so that it allows large price jumps if seen repeatedly
+  if (
+    priceSoFar.gt(token.lastPriceUSD!.times(BIGDECIMAL_TWO)) ||
+    priceSoFar.lt(token.lastPriceUSD!.div(BIGDECIMAL_TWO))
+  ) {
+    if (token._largePriceChangeBuffer < PRICE_CHANGE_BUFFER_LIMIT) {
+      token._largePriceChangeBuffer += 1;
+      token.save();
+      return token.lastPriceUSD!;
+    }
+  }
+
+  token._largePriceChangeBuffer = 0;
+  token._largeTVLImpactBuffer = 0;
+
+  token.save();
+  return priceSoFar;
+}
+
+// Tried to return null from here and it did not
+function get_token_index(pool: LiquidityPool, token: Token): i32 {
+  if (pool.inputTokens[0] == token.id) {
+    return 0;
+  }
+  if (pool.inputTokens[1] == token.id) {
+    return 1;
+  }
+  return -1;
+}
+
+function isValidTVL(
+  pool: LiquidityPool,
+  token: Token,
+  price: BigDecimal
+): bool {
+  const newTokenTotalValueLocked = convertTokenToDecimal(
+    token._totalSupply,
+    token.decimals
+  ).times(price);
+
+  // If price is too high, skip this pool
+  if (newTokenTotalValueLocked.gt(BIGDECIMAL_TEN_BILLION)) {
+    log.warning("Price too high for token: {} from pool: {}", [
```

Comment:
probably an error log

Reply:
https://github.com/messari/subgraphs/blob/master/docs/WALKTHROUGH.md#using-logs-in-subgraphs

Reply:
warning seems correct to me in this case

Reply:
The shared uniswap fork code uses warning, so I agree to stick w the precedent here